src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -258,97 +296,162 @@ void q_omg_log_exception(const struct ddsrt_log_cfg *lc, uint32_t cat, DDS_Secur
-static struct security_entity_match * remove_entity_match(struct dds_security_match_index *list, const ddsi_guid_t *src, const ddsi_guid_t *dst)
+static void clear_pending_matches_by_local_guid(dds_security_context *sc, struct pending_match_index *index, const ddsi_guid_t *local_guid)
 {
-  struct security_entity_match *match;
-  struct guid_pair guids;
-  ddsrt_avl_dpath_t path;
-
-  guids.src = *src;
-  guids.dst = *dst;
+  struct pending_match *match;
 
-  ddsrt_mutex_lock(&list->lock);
-  match = ddsrt_avl_lookup_dpath(&entity_match_treedef, &list->matches, &guids, &path);
-  if (match)
-    ddsrt_avl_delete_dpath(&entity_match_treedef, &list->matches, match, &path);
-  ddsrt_mutex_unlock(&list->lock);
-
-  return match;
+  ddsrt_mutex_lock(&index->lock);
+  match = ddsrt_avl_find_min(&pending_match_index_treedef, &index->pending_matches);
+  while (match)
+  {
+    struct pending_match *next = ddsrt_avl_find_succ(&pending_match_index_treedef, &index->pending_matches, match);
+    if (guid_compare(&match->guids.local_guid, local_guid) == 0)
+    {
+      ddsrt_avl_delete(&pending_match_index_treedef, &index->pending_matches, match);
+      clear_and_free_pending_match(sc, match);
+    }
+    match = next;
+  }
+  ddsrt_mutex_unlock(&index->lock);
 }
 
-static struct dds_security_match_index * security_match_index_new(void)
+static void clear_pending_matches_by_remote_guid(dds_security_context *sc, struct pending_match_index *index, const ddsi_guid_t *remote_guid)
 {
-  struct dds_security_match_index *list;
+  struct guid_pair template = { .remote_guid = *remote_guid, .local_guid = {.prefix.u = {0, 0, 0}, .entityid.u = 0} };
+  struct pending_match *match;
 
-  list  = ddsrt_malloc (sizeof(*list));
-  ddsrt_mutex_init (&list->lock);
-  ddsrt_avl_init (&entity_match_treedef, &list->matches);
-  return list;
+  ddsrt_mutex_lock(&index->lock);
+  match = ddsrt_avl_lookup_succ(&pending_match_index_treedef, &index->pending_matches, &template);
+  while (match && guid_compare(&match->guids.remote_guid, remote_guid) == 0)
+  {
+    struct pending_match *next = ddsrt_avl_lookup_succ(&pending_match_index_treedef, &index->pending_matches, &match->guids);
+    ddsrt_avl_delete(&pending_match_index_treedef, &index->pending_matches, match);
+    clear_and_free_pending_match(sc, match);
+    match = next;
+  }
+  ddsrt_mutex_unlock(&index->lock);
 }
 
-static void entity_match_free_wrapper(void *arg)
+static void pending_match_index_init(const struct ddsi_domaingv *gv, struct pending_match_index *index)
 {
-  struct security_entity_match *match = arg;
-  entity_match_free(match);
+  ddsrt_mutex_init(&index->lock);
+  ddsrt_avl_init(&pending_match_index_treedef, &index->pending_matches);
+  ddsrt_fibheap_init(&pending_match_expiry_fhdef, &index->expiry_timers);
+  index->gv = gv;
+  index->evt = NULL;
 }
 
-static void security_match_index_free(struct dds_security_match_index *list)
+static void pending_match_index_deinit(struct pending_match_index *index)
```

Comment:
Had to add an extra deinit function of the security context to implement this. This because the security context can only be freed when the xevent queue has been freed before. Callbacks scheduled on the xevent queue could still access the security context.